### PR TITLE
Add unit tests in some push classes

### PR DESCRIPTION
**Pull request by bmarty**
Add unit tests in some push classes

I had to introduce some interface to be able to properly unit test more classes.

Code coverage is doing a nice bump:

image

This does not fix any bug, except one related to log, so very minor.

### CODE REVIEW

File: libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/push/OnNotifiableEventReceived.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import io.element.android.libraries.push.impl.notifications.DefaultNotificationDrawerManager
+import io.element.android.libraries.push.impl.notifications.model.NotifiableEvent
+import javax.inject.Inject
+
+interface OnNotifiableEventReceived {
+    fun onNotifiableEventReceived(notifiableEvent: NotifiableEvent)
+}
+
+@ContributesBinding(AppScope::class)
+class DefaultOnNotifiableEventReceived @Inject constructor(
```
Comment (jmartinesp) [inline libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/push/OnNotifiableEventReceived.kt:30]: We could move this to its own file, maybe.

File: libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/push/OnNotifiableEventReceived.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+package io.element.android.libraries.push.impl.push
+
+import com.squareup.anvil.annotations.ContributesBinding
+import io.element.android.libraries.di.AppScope
+import io.element.android.libraries.push.impl.notifications.DefaultNotificationDrawerManager
+import io.element.android.libraries.push.impl.notifications.model.NotifiableEvent
+import javax.inject.Inject
+
+interface OnNotifiableEventReceived {
```
Comment (jmartinesp) [inline libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/push/OnNotifiableEventReceived.kt:25]: Docs?

File: libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/push/IncrementPushDataStore.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package io.element.android.libraries.push.impl.push
+
+import com.squareup.anvil.annotations.ContributesBinding
+import io.element.android.libraries.di.AppScope
+import io.element.android.libraries.push.impl.store.DefaultPushDataStore
+import javax.inject.Inject
+
+interface IncrementPushDataStore {
```
Comment (jmartinesp) [inline libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/push/IncrementPushDataStore.kt:24]: Docs?

File: libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/push/IncrementPushDataStore.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    suspend fun incrementPushCounter()
+}
+
+@ContributesBinding(AppScope::class)
+class DefaultIncrementPushDataStore @Inject constructor(
```
Comment (jmartinesp) [inline libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/push/IncrementPushDataStore.kt:29]: As with the other interface + default implementations, maybe we should split them in 2 files?

File: libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/pushgateway/PushGatewayApiFactory.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package io.element.android.libraries.push.impl.pushgateway
+
+import com.squareup.anvil.annotations.ContributesBinding
+import io.element.android.libraries.di.AppScope
+import io.element.android.libraries.network.RetrofitFactory
+import javax.inject.Inject
+
+interface PushGatewayApiFactory {
```
Comment (jmartinesp) [inline libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/pushgateway/PushGatewayApiFactory.kt:24]: Docs + splitting?

File: libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/pushgateway/PushGatewayNotifyRequest.kt [hunk 6]
```diff
@@ -15,15 +15,14 @@
...
 import io.element.android.libraries.matrix.api.core.EventId
 import io.element.android.libraries.matrix.api.core.RoomId
-import io.element.android.libraries.network.RetrofitFactory
 import io.element.android.libraries.push.api.gateway.PushGatewayFailure
 import javax.inject.Inject
 
-class PushGatewayNotifyRequest @Inject constructor(
-    private val retrofitFactory: RetrofitFactory,
-) {
+interface PushGatewayNotifyRequest {
```
Comment (jmartinesp) [inline libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/pushgateway/PushGatewayNotifyRequest.kt:25]: Docs + splitting?

File: libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/test/TestPush.kt [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import com.squareup.anvil.annotations.ContributesBinding
+import io.element.android.appconfig.PushConfig
+import io.element.android.libraries.di.AppScope
+import io.element.android.libraries.matrix.api.core.EventId
+import io.element.android.libraries.matrix.api.core.RoomId
+import io.element.android.libraries.push.impl.pushgateway.PushGatewayNotifyRequest
+import io.element.android.libraries.pushproviders.api.CurrentUserPushConfig
+import javax.inject.Inject
+
+interface TestPush {
```
Comment (jmartinesp) [inline libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/test/TestPush.kt:28]: Docs + splitting?

File: libraries/pushproviders/firebase/src/main/kotlin/io/element/android/libraries/pushproviders/firebase/FirebaseNewTokenHandler.kt [hunk 8]
```diff
@@ -32,14 +34,19 @@ private val loggerTag = LoggerTag("FirebaseNewTokenHandler", LoggerTag.PushLogge
 /**
  * Handle new token receive from Firebase. Will update all the sessions which are using Firebase as a push provider.
  */
-class FirebaseNewTokenHandler @Inject constructor(
+interface FirebaseNewTokenHandler {
```
Comment (jmartinesp) [inline libraries/pushproviders/firebase/src/main/kotlin/io/element/android/libraries/pushproviders/firebase/FirebaseNewTokenHandler.kt:37]: Split into 2 files?

File: libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/RegisterUnifiedPushUseCase.kt [hunk 9]
```diff
@@ -17,43 +17,41 @@
...
 import kotlinx.coroutines.flow.filter
 import kotlinx.coroutines.flow.first
-import kotlinx.coroutines.launch
 import kotlinx.coroutines.withTimeout
 import org.unifiedpush.android.connector.UnifiedPush
 import javax.inject.Inject
 import kotlin.time.Duration.Companion.seconds
 
-class RegisterUnifiedPushUseCase @Inject constructor(
+interface RegisterUnifiedPushUseCase {
```
Comment (jmartinesp) [inline libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/RegisterUnifiedPushUseCase.kt:32]: Docs + splitting?

File: libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnifiedPushApiFactory.kt [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+package io.element.android.libraries.pushproviders.unifiedpush
+
+import com.squareup.anvil.annotations.ContributesBinding
+import io.element.android.libraries.di.AppScope
+import io.element.android.libraries.network.RetrofitFactory
+import io.element.android.libraries.pushproviders.unifiedpush.network.UnifiedPushApi
+import javax.inject.Inject
+
+interface UnifiedPushApiFactory {
```
Comment (jmartinesp) [inline libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnifiedPushApiFactory.kt:25]: Docs + splitting?

File: libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnifiedPushGatewayResolver.kt [hunk 11]
```diff
@@ -16,29 +16,33 @@
...
-import io.element.android.libraries.pushproviders.unifiedpush.network.UnifiedPushApi
+import io.element.android.libraries.di.AppScope
 import kotlinx.coroutines.withContext
 import timber.log.Timber
 import java.net.URL
 import javax.inject.Inject
 
-class UnifiedPushGatewayResolver @Inject constructor(
-    private val retrofitFactory: RetrofitFactory,
+interface UnifiedPushGatewayResolver {
```
Comment (jmartinesp) [inline libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnifiedPushGatewayResolver.kt:27]: Docs + splitting?

File: libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnifiedPushGatewayResolver.kt [hunk 12]
```diff
@@ -16,29 +16,33 @@
...
-        val port = if (url.port != -1) ":${url.port}" else ""
-        val customBase = "${url.protocol}://${url.host}$port"
-        val customUrl = "$customBase/_matrix/push/v1/notify"
-        Timber.i("Testing $customUrl")
         try {
+            val url = URL(endpoint)
+            val port = if (url.port != -1) ":${url.port}" else ""
+            val customBase = "${url.protocol}://${url.host}$port"
+            val customUrl = "$customBase/_matrix/push/v1/notify"
+            Timber.i("Testing $customUrl")
```
Comment (jmartinesp) [inline libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnifiedPushGatewayResolver.kt:43]: Should we re-throw `CancellationExceptions`?

File: libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnifiedPushNewGatewayHandler.kt [hunk 13]
```diff
@@ -16,27 +16,34 @@
...
 import javax.inject.Inject
 
-private val loggerTag = LoggerTag("UnifiedPushNewGatewayHandler", LoggerTag.PushLoggerTag)
+private val loggerTag = LoggerTag("DefaultUnifiedPushNewGatewayHandler", LoggerTag.PushLoggerTag)
 
 /**
  * Handle new endpoint received from UnifiedPush. Will update the session matching the client secret.
  */
-class UnifiedPushNewGatewayHandler @Inject constructor(
+interface UnifiedPushNewGatewayHandler {
```
Comment (jmartinesp) [inline libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnifiedPushNewGatewayHandler.kt:35]: Split this into 2 files?

File: libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnifiedPushStore.kt [hunk 14]
```diff
@@ -19,32 +19,44 @@ package io.element.android.libraries.pushproviders.unifiedpush
...
 import androidx.core.content.edit
+import com.squareup.anvil.annotations.ContributesBinding
+import io.element.android.libraries.di.AppScope
 import io.element.android.libraries.di.ApplicationContext
 import io.element.android.libraries.di.DefaultPreferences
 import io.element.android.libraries.matrix.api.core.UserId
 import javax.inject.Inject
 
-class UnifiedPushStore @Inject constructor(
+interface UnifiedPushStore {
```
Comment (jmartinesp) [inline libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnifiedPushStore.kt:29]: Docs + splitting?

File: libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnregisterUnifiedPushUseCase.kt [hunk 15]
```diff
@@ -17,27 +17,34 @@
...
 import io.element.android.libraries.pushproviders.api.PusherSubscriber
 import org.unifiedpush.android.connector.UnifiedPush
 import javax.inject.Inject
 
-class UnregisterUnifiedPushUseCase @Inject constructor(
+interface UnregisterUnifiedPushUseCase {
+    suspend fun execute(matrixClient: MatrixClient, clientSecret: String): Result<Unit>
+}
+
+@ContributesBinding(AppScope::class)
```
Comment (jmartinesp) [inline libraries/pushproviders/unifiedpush/src/main/kotlin/io/element/android/libraries/pushproviders/unifiedpush/UnregisterUnifiedPushUseCase.kt:32]: Docs + splitting?

File: libraries/textcomposer/impl/src/main/kotlin/io/element/android/libraries/textcomposer/mentions/MentionSpanProvider.kt [hunk 16]
```diff
@@ -164,7 +164,7 @@ internal fun MentionSpanPreview() {
                         eventId = null,
                         viaParameters = persistentListOf(),
                     )
-                    else -> TODO()
+                    else -> throw AssertionError("Unexpected value $uriString")
```
Comment (jmartinesp) [inline libraries/textcomposer/impl/src/main/kotlin/io/element/android/libraries/textcomposer/mentions/MentionSpanProvider.kt:167]: 👍

**Review by jmartinesp — APPROVED** [review]
Thanks for the work on this and the big coverage increase! I added lots of small comments (especially about docs), but none of them are really blockers, so feel free to merge it as is.

**Comment by bmarty** [comment]
Thanks for the review. About splitting interface and default implementation, I think it does not worth it as the interface in only to ease the unit tests. so it's actually fine to me if the interface (which does not need to be expose in a api module) and the default impl is in the same file.
Let's merge the PR like this since as you said there is no blocker.
Thanks again for the review!

**Pull request merged**